src/tls/test/main.cpp

```diff
@@ -0,0 +1,378 @@
+  auto pk = kp->private_key_pem();
+  return make_unique<Cert>(move(ca), crt, pk);
+}
+
+/// Test runner, with various options for different kinds of tests.
+void run_test_case(
+  int dgram,
+  const uint8_t* message,
+  size_t message_length,
+  const uint8_t* response,
+  size_t response_length,
+  unique_ptr<tls::Cert> server_cert,
+  unique_ptr<tls::Cert> client_cert,
+  bool requires_auth)
+{
+  uint8_t buf[max(message_length, response_length) + 1];
+
+  // Create a pair of client/server
+  tls::Server server(move(server_cert), dgram);
+  server.set_require_auth(requires_auth);
+  tls::Client client(move(client_cert), dgram);
+  client.set_require_auth(requires_auth);
+
+  // Connect BIOs together
+  TestPipe pipe(dgram);
+  server.set_bio(
+    &pipe, send<TestPipe::SERVER>, recv<TestPipe::SERVER>, dbg_callback);
+  client.set_bio(
+    &pipe, send<TestPipe::CLIENT>, recv<TestPipe::CLIENT>, dbg_callback);
+
+  // There could be multiple communications between client/server while
+  // doing the handshake, and they won't return an error until there's
+  // nothing more to read/write, so we put them in separate threads.
+
+  // Create a thread for the client handshake
+  thread client_thread([&client]() {
+    test_log.write("Client handshake\n");
+    if (handshake(&client))
+      throw runtime_error("Client handshake error\n");
+  });
+
+  // Create a thread for the server handshake
+  thread server_thread([&server]() {
+    test_log.write("Server handshake\n");
+    if (handshake(&server))
+      throw runtime_error("Client handshake error\n");
+  });
+
+  // Join threads
+  client_thread.join();
+  server_thread.join();
+  test_log.write("Handshake completed\n");
+
+  // The rest of the communication is deterministic and easy to simulate
+  // so we take them out of the thread, to guarantee there will be bytes
+  // to read at the right time.
+  if (message_length == 0)
+  {
+    test_log.write("Empty message. Ignoring communication test\n");
+    test_log.write("Closing connection\n");
+    client.close();
+    server.close();
+    return;
+  }
+
+  // Send the first message
+  test_log.write(
+    fmt::format("Client sending message [{}]\n", string((const char*)message)));
+  int written = client.write(message, message_length);
+  REQUIRE(written == message_length);
+
+  // Receive the first message
+  int read = server.read(buf, message_length);
+  REQUIRE(read == message_length);
+  buf[message_length] = '\0';
+  test_log.write(
+    fmt::format("Server message received [{}]\n", string((const char*)buf)));
+  REQUIRE(strncmp((const char*)buf, (const char*)message, message_length) == 0);
+
+  // Send the response
+  test_log.write(fmt::format(
+    "Server sending message [{}]\n", string((const char*)response)));
+  written = server.write(response, response_length);
+  REQUIRE(written == response_length);
+
+  // Receive the response
+  read = client.read(buf, response_length);
```

Comment:
Can we add a test that sends a long message, larger than the crypto implementation will process in a single read? I believe `mbedtls` has some internal buffer and will only give us (iirc) 4k on each call to read, so there are cases where we need to call `client.read` multiple times to retrieve everything from the corresponding `write`.